feat: add webperf-core-web-vitals skill (#1198)
ChromeDevTools/chrome-devtools-mcp: skills/webperf-core-web-vitals/SKILL.md

context: fork
---

# WebPerf: Core Web Vitals

JavaScript snippets for measuring web performance in Chrome DevTools. Execute with `evaluate_script`, capture output with `get_console_message`.

## Scripts

- `scripts/CLS.js` — Cumulative Layout Shift (CLS)
- `scripts/INP.js` — Interaction to Next Paint (INP)
- `scripts/LCP-Image-Entropy.js` — LCP Image Entropy
- `scripts/LCP-Sub-Parts.js` — LCP Sub-Parts
- `scripts/LCP-Trail.js` — LCP Trail
- `scripts/LCP-Video-Candidate.js` — LCP Video Candidate
- `scripts/LCP.js` — Largest Contentful Paint (LCP)

Descriptions, thresholds, and return schemas: `references/snippets.md`, `references/schema.md`

## Script Execution Patterns

Scripts fall into two execution patterns:

### Synchronous (LCP, CLS, LCP-Sub-Parts, LCP-Trail, LCP-Image-Entropy, LCP-Video-Candidate)

Run via `evaluate_script` and return structured JSON immediately from buffered performance data. The page must have already loaded.

> ### Tracking (INP)

Reviewer: ```suggestion
### Measuring (INP)
```

INP requires real user interactions to measure. The workflow is:

1. Run `INP.js` via `evaluate_script` → returns `{ status: "tracking", getDataFn: "getINP" }`

> 2. **Tell the user:** "INP tracking is now active. Please interact with the page — click buttons, open menus, fill form fields — then let me know when you're done."

Reviewer: ```suggestion
1. Run `INP.js` via `evaluate_script` → returns `{ status: "measuring", getDataFn: "getINP" }`
2. **Tell the user:** "INP measuring is now active. Please interact with the page — click buttons, open menus, fill form fields — then let me know when you're done."
```

3. Wait for the user to confirm they've interacted.
4. Call `evaluate_script("getINP()")` to collect results.
5. If `getINP()` returns `status: "error"` → the user has not interacted yet. Remind them and wait.
6. For a full breakdown of all interactions, call `evaluate_script("getINPDetails()")` — returns all recorded interactions sorted by duration.

> The agent cannot interact with the page on behalf of the user for INP measurement. Real user interactions are required.

## Common Workflows

### Complete Core Web Vitals Audit

When the user asks for a comprehensive Core Web Vitals analysis or "audit CWV":

1. **LCP.js** - Measure Largest Contentful Paint
2. **CLS.js** - Measure Cumulative Layout Shift
3. **INP.js** - Measure Interaction to Next Paint

> 4. **LCP-Sub-Parts.js** - Break down LCP timing phases

Reviewer: ```suggestion
4. **LCP-Subparts.js** - Break down LCP timing phases
```

5. **LCP-Trail.js** - Track LCP candidate evolution

### LCP Deep Dive

When LCP is slow or the user asks "debug LCP" or "why is LCP slow":

1. **LCP.js** - Establish baseline LCP value

> 2. **LCP-Sub-Parts.js** - Break down into TTFB, resource load, render delay

Reviewer: ```suggestion
2. **LCP-Subparts.js** - Break down into TTFB, resource load, render delay
```

3. **LCP-Trail.js** - Identify all LCP candidates and changes
4. **LCP-Image-Entropy.js** - Check if LCP image has visual complexity issues
5. **LCP-Video-Candidate.js** - Detect if LCP is a video (poster or video element)

### CLS Investigation

When layout shifts are detected or the user asks "debug CLS" or "layout shift issues":

1. **CLS.js** - Measure overall CLS score
2. **Layout-Shift-Loading-and-Interaction.js** (from `webperf-interaction` skill) - Separate loading vs interaction shifts
3. Cross-reference with **webperf-loading** skill:
   - Find-Above-The-Fold-Lazy-Loaded-Images.js (lazy images causing shifts)
   - Fonts-Preloaded-Loaded-and-used-above-the-fold.js (font swap causing shifts)

### INP Debugging

When interactions feel slow or the user asks "debug INP" or "slow interactions":

> 1. **INP.js** - Start tracking. Tell the user to interact with the page and confirm when done.

Reviewer: ```suggestion
1. **INP.js** - Start measuring. Tell the user to interact with the page and confirm when done.
```

2. Call `getINP()` to collect results once the user confirms.
3. Call `getINPDetails()` to see all interactions ranked by duration.
4. **Interactions.js** (from `webperf-interaction` skill) - List all interactions with timing
5. **Input-Latency-Breakdown.js** (from `webperf-interaction` skill) - Break down input delay, processing, presentation
6. **Long-Animation-Frames.js** (from `webperf-interaction` skill) - Identify blocking animation frames
7. **Long-Animation-Frames-Script-Attribution.js** (from `webperf-interaction` skill) - Find scripts causing delays

### Video as LCP Investigation

When LCP is a video element (detected by LCP-Video-Candidate.js):

1. **LCP-Video-Candidate.js** - Identify video as LCP candidate
2. **Video-Element-Audit.js** (from Media skill) - Audit video loading strategy

> 3. **LCP-Sub-Parts.js** - Analyze video loading phases

Reviewer: ```suggestion
3. **LCP-Subparts.js** - Analyze video loading phases
```

4. Cross-reference with **webperf-loading** skill:
   - Resource-Hints-Validation.js (check for video preload)

>    - Priority-Hints-Audit.js (check for fetchpriority on video)

Reviewer: Should these be here? Ah I see these are part of https://github.com/ChromeDevTools/chrome-devtools-mcp/issues/1114

### Image as LCP Investigation

When LCP is an image (most common case):

1. **LCP.js** - Measure LCP timing

> 2. **LCP-Sub-Parts.js** - Break down timing phases

Reviewer: ```suggestion
2. **LCP-Subparts.js** - Break down timing phases
```

3. **LCP-Image-Entropy.js** - Analyze image complexity
4. Cross-reference with **webperf-media** skill:
   - Image-Element-Audit.js (check format, dimensions, lazy loading)
5. Cross-reference with **webperf-loading** skill:
   - Find-Above-The-Fold-Lazy-Loaded-Images.js (check if incorrectly lazy)
   - Priority-Hints-Audit.js (check for fetchpriority="high")

>    - Resource-Hints-Validation.js (check for preload)

Reviewer: Needs https://github.com/ChromeDevTools/chrome-devtools-mcp/issues/1114 to land first

## Decision Tree

Use this decision tree to automatically run follow-up snippets based on results:

### After LCP.js

- **If LCP > 2.5s** → Run **LCP-Sub-Parts.js** to diagnose which phase is slow
- **If LCP > 4.0s (poor)** → Run full LCP deep dive workflow (5 snippets)
- **If LCP candidate is an image** → Run **LCP-Image-Entropy.js** and **webperf-media:Image-Element-Audit.js**
- **If LCP candidate is a video** → Run **LCP-Video-Candidate.js** and **webperf-media:Video-Element-Audit.js**
- **Always run** → **LCP-Trail.js** to understand candidate evolution

> ### After LCP-Sub-Parts.js

Reviewer: ```suggestion
### After LCP-Subparts.js
```
